### ISSUE TITLE
Deadlock while running scrub

### ISSUE THREAD

**Issue by Gibson85** [openzfs/zfs#15815]
Hi, revently I had an unfixable error on a file of my ZFS Raid 1. I wonder how this could happen. But ok. I deleted this file mentioned under zpool status and ran a scrub to check for more issues. The scrub run stucked at some point. Also I could hear the drives are not used anymore and the expected time to finish the scrub was increasing. Thus I wanted to stop the scrub but this (scrub -s) command also stucked and never finished. Also the PC could not shutdown anymore and I had to power it off. A second scrub almost finished but I had to stop it because I went to vaction then. Now I started a new scrub and it was stuck again this morning at 30%. First time I powered of the PC there was the message:

Please check any tasks stuck between calls to exit_tasks_rcu_start() and exit_tasks_rcu_finish()

tasks_rcu_exit_stall: rcu_tasks grace period number 41 (since boot) gp_state: RTGS_POST_SCAN_TASKLIST

This maybe is only a follow up error. I also disconnect my USB HUB and router when running a scrub to safe energy because it takes a day to finish the scrub.

It's a shame. ZFS can't do it's basic tasks anymore.

Also you still haven't fixed the bug (Fedora) that with every kernel update ZFS is broken and one needs to dnf remove and dnf install it again to make it work.

Seriously. It looks you have lost completely the control over your codebase. Only have a look at all the issues for years.

**Comment by rincebrain** [comment]
You deleted the entire issue template and most of this is just a rant.

You didn't include dmesg from when it was stuck, or the version of OpenZFS you were using, or the distro.

I would recommend you try including that information if you want any constructive replies and not just to rant - and if you do just want to rant, a bug report is not really the correct forum for it.

**Comment by Gibson85** [comment]
As mentioned above I am using fedora with the actual ZFS version. I did a new scrub this night being stuck again this morning at 40%. So this behaviour is reproducable. Beside that almost no application can be started anymore. The hole system seems stuck. The dmesg file I will provide later.

**Comment by Gibson85** [comment]
Please find attached the dmesg file.
[dmesg.txt](https://github.com/openzfs/zfs/files/14047968/dmesg.txt)

**Comment by gmelikov** [comment]
> reporting a critical bug

Without issue template you report something, but not issue this project is waiting for.

> I am using fedora with the actual ZFS version

What is "actual" version? 2.1.X/2.2.X/master? Even 2.2.X may have new version in a day-week-month, it's not constructive at all.

> And giving thumbs down to someone reporting a critical bug shows your arrogance.

Honestly, your "It looks you have lost completely the control over your codebase" is not reporting a critical bug, it's your emotions, which demotivate anybody, so people give you their emotions back with reactions. It's an opensource community, not a paid proprietary support.

But let's be more constructive, for now it's a report only from you, so looks like it may be a problem with hardware. Do you have other problems, or only with scrub? Could you read/write data on pool? Does it reproduce on other hardware? Unfortunately, if you can reproduce something on specific hardware, OS and pool, it doesn't guarantee that it could be reproduced by others. If you didn't found existing issues with this problem - then now only you can reproduce it (community have active people with latest fedora). 

So, we need to bisect the problem from your side.

**Comment by Gibson85** [comment]
It's ZFS 2.2.2-1, ZFS pool version 5000, ZFS filesystem version 5. After reboot this morning the scrub was continuing automatically but had fallrn back from around 40% to 30%. I let it continue and now it is stuck at 59% again. So at least with my maschine it is reproducable. If you need further informations let me know. I hope a stucked scrub cannot lead to corrupt data?

**Comment by Gibson85** [comment]
What about lines 1313 to 1346 of my dmesg file above. I am not an expert. But the word "RIP" normally means nothing good I guess. Maybe a problem with a AVL tree rotation?

**Comment by snajpa** [comment]
While also not a fan of your style or attitude, I somewhat agree there are more bugs lately. But this is not the way to deal with the situation, trust me :D It isn't for the better. There's only a handful of full-time people working on this codebase, most of it is done on people's free time, where they are the ones in charge of their own time. This attitude tends to move the scales to the "let's do something else now" direction, which is not what we want, is it? :)

I think this issue https://github.com/openzfs/zfs/issues/15271 seems to be relevant, only on the FreeBSD side (but this is a common code path)

More info as per the issue can be found here https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=268909

**Comment by Gibson85** [comment]
Ok, I think I found the reason for this behaviour. Getting corrupted file entries at ZFS status last days after copying files (e.g. for backup) I checked all my hardware and found out that one of my RAM modules became faulty. After replacing the CPU and RAM (with ECC now) everything seems fine. Also the last two scrubs did not get stuck anymore. Also there had been multiple kernel updates last days. Maybe there was fixed something also. Thus I close this ticket. If I experience some problems again I will make a new ticket. Also I will do a fresh fedora install soon (if the newly kernel 6.7.3 problems with NVIDIA/AMD cards are fixed...). I will see then if the problem that with every kernel update ZFS is broken (requires uninstall/install) will reappear again.

So sorry for my harsh writing but when it comes to dataloss I don't understand fun. Luckily after changing the RAM and making a new scrub again the files ZFS had marked as corrupted had been cleared out automatically. So these had not been corrupted at all. Finally ZFS did a good job here.

So lesson learend. In future I will do a monthly RAM check + scrub.
